core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,225 @@
+
+  /**
+   * Called when this iterator is on the latest element by `hasNext`.
+   * This method will rename temporary output path to final output path of checkpoint data.
+   */
+  private[this] def complete(): Unit = {
+    if (completed) {
+      return
+    }
+
+    if (serializeStream == null) {
+      // There is some exception when creating serializeStream, we only need to clean up the
+      // resources.
+      cleanup()
+      return
+    }
+
+    while (values.hasNext) {
+      serializeStream.writeObject(values.next)
+    }
+
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    serializeStream.close()
+
+    if (!fs.rename(tempOutputPath, finalOutputPath)) {
+      if (!fs.exists(finalOutputPath)) {
+        logInfo("Deleting tempOutputPath " + tempOutputPath)
+        fs.delete(tempOutputPath, false)
+        throw new IOException("Checkpoint failed: failed to save output of task: "
+          + context.attemptNumber + " and final output path does not exist")
+      } else {
+        // Some other copy of this task must've finished before us and renamed it
+        logInfo("Final output path " + finalOutputPath + " already exists; not overwriting it")
+        fs.delete(tempOutputPath, false)
+      }
+    }
+  }
+
+  private[this] def cleanup(): Unit = {
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    if (serializeStream != null) {
+      serializeStream.close()
+    }
+    fs.delete(tempOutputPath, false)
+  }
+
+  override def hasNext: Boolean = {
+    try {
+      val r = values.hasNext
+      if (!r) {
+        complete()
+      }
+      r
+    } catch {
+      case e: Throwable =>
+        try {
+          cleanup()
+        } catch {
+          case NonFatal(e1) =>
+            // Log `e1` since we should not override `e`
+            logError(e1.getMessage, e1)
+        }
+        throw e
+    }
+  }
+
+  override def next(): T = {
+    try {
+      val value = values.next()
+      serializeStream.writeObject(value)
+      value
+    } catch {
+      case e: Throwable =>
+        try {
+          cleanup()
+        } catch {
+          case NonFatal(e1) =>
+            // Log `e1` since we should not override `e`
+            logError(e1.getMessage, e1)
+        }
+        throw e
+    }
+  }
+}
+
+private[spark] object CheckpointingIterator {
+
+  private val checkpointingRDDPartitions = new ConcurrentHashMap[RDDBlockId, RDDBlockId]()
+
+  /**
+   * Return true if the caller gets the lock to write the checkpoint file. Otherwise, the caller
+   * should not do checkpointing.
+   */
+  private def acquireLockForPartition(id: RDDBlockId): Boolean = {
+    checkpointingRDDPartitions.putIfAbsent(id, id) == null
+  }
+
+  /**
+   * Release the lock to avoid memory leak.
+   */
+  private def releaseLockForPartition(id: RDDBlockId): Unit = {
+    checkpointingRDDPartitions.remove(id)
+  }
+
+  def apply[T: ClassTag](
```

Comment:
need big java doc to explain what it's doing. We should mention the implicit locking this does and why it's needed etc.